pkg/server/plugin/zmq/rpc.go

```diff
@@ -176,6 +194,54 @@ func (p *zmqTransport) ipc(req *pluginrequest.Request) (out []byte, err error) {
 	return
 }
 
+func (p *zmqTransport) handleRequest(parcel *Parcel) {
+	workerID := parcel.Worker
+	requestID := parcel.RequestID
+	bounceCount := parcel.BounceCount
+	buffer := parcel.Frame
+	responseChan := parcel.RespChan
+
+	reader := bytes.NewReader(buffer)
+	data := map[string]interface{}{}
+	if jsonErr := json.NewDecoder(reader).Decode(&data); jsonErr != nil && jsonErr != io.EOF {
+		// TODO(b123400): handle json error
+		return
+	}
+
+	payloadData := data["payload"].(map[string]interface{})
+	method := data["method"].(string)
+	path := data["path"].(string)
+
+	payload := &router.Payload{
+		Context: context.Background(),
+		Meta: map[string]interface{}{},
+		Data: payloadData,
+		AccessKey: router.MasterAccessKey,
+	}
+
+	responseWriter := &zmqResponseWriter{}
+	resp := router.NewResponse(responseWriter)
+
+	newCtx := context.WithValue(payload.Context, ZMQWorkerIDContextKey, workerID)
+	newCtx = context.WithValue(newCtx, ZMQRequestIDContextKey, requestID)
+	newCtx = context.WithValue(newCtx, ZMQBounceCountContextKey, bounceCount)
+	payload.Context = newCtx
+
+	// TODO(b123400), put real path and method here
+	p.router.HandlePayload(path, method, payload, resp)
+
+	responseChan <- responseWriter.response
+}
+
+func (p *zmqTransport) listenRequests() {
+	for {
+		select {
+		case parcel := <- p.broker.ReqChan:
+			go p.handleRequest(parcel)
```

Comment:
Um... let me confirm if I understand your comment correctly. The current implementation is
1. Broker puts parcel to `ReqChan`
2. zmqTransport listens `ReqChan`, and spawn a thread (handleRequest) per request
3. `handleRequest` does things synchronously 

So it is `broker -> listenRequests -> handleRequest`, 3 levels of async.
Your suggestion is, `broker -> handleRequest`, 2 level of async.

When I write the code, I am not really concerning about how many levels, or is it too small or big, but their responsibility, their scope.
Broker has been focusing on 1 thing only, sending and receiving zmq messages. Both input and output of broker has always been just buffer, it receives buffers via `.RPC` and output via `reqChan`. It does not have the knowledge of transport (the Transport interface). It doesn't even care if the response is handled, if you don't care about the response, you can just throw away the channel and everything will be fine.
On the other hand, `zmqTransport` is more about application logic, it transform concept of Request and Response into buffer, bridging the abstract concept of request response to the underlying bytes.
`zmqTransport` has the knowledge of the broker, but not the other way round. That is what I am try to maintain while making the broker interface, broker should output a stream of requests, it will be callback if it is Javascript, or delegate if it is Objective-C. 

If there is a single piece of code, somewhere, doing both `make_parcel()` and `doRequest(parcel)`, we will be crossing the line, now broker will know about the transport, it will enforce every request has to be handled. While it is always the case that requests are handled, the 2 pieces of code will be tied together and we are not separating the responsibilities anymore.

Speaking of Erlang, I think it's kind of different because their threading is more lightweight, and they do encourage many small processes.